src/group-by.mjs

```diff
@@ -1,39 +1,97 @@
 import { iterableCurry } from './internal/iterable'
+import { Exchange } from './internal/queues'
 
-function * groupBy (key = (k) => k, iterable) {
+const UNIQUE_INITIAL_KEY = {}
+
+function groupBy (getKey = (k) => k, iterable) {
   const iterator = iterable[Symbol.iterator]()
```

Comment:
This is a bug actually. Never noticed it before. We can't open the source iterator until we're inside the try block with the finally that cleans it up.

Reply:
I don't think this is a problem.
If you have never called "yield" on the iterator, you won't need to clean it up.

Reply:
That's not true. It is true of our iterators, but not of iterators or iterables in general.

Reply:
can you explain better?
```js
const iterator = iterable[Symbol.iterator]()
```
should return an iterator or an exception and they are both fine.

You need to call
```js
iterator.return()
```
only after the first call to
```js
iterator.next()
```

Reply:
Remember iterators aren't always generator functions. Someone could give you an iterable over a file:
```js
class FileIterator {
  constructor(name) {
    this.handle = fs.openSync(name)
  }
  next() { /* yield line from file */
  return() {
    fs.closeSync(this.handle)
  }
}

class File {
  [Symbol.iterator]() {
    return new FileIterator(this.name)
  }
}

Array.from(new File('./myFile'))
```

Reply:
So what you say about needing to call `return` only after the first call to `next` really isn't true. It is usually OK in our code because we usually open external iterators on our first iteration, not before.

Reply:
That would be an incorrect use of the iterator interface. And it won't work correctly in a for..of loop either.

Reply:
If you believe that would be an incorrect use of the iterator interface, could you point me to where it is specified that that is not allowed?

Reply:
https://tc39.github.io/ecma262/#sec-iterator-interface

> The returned object must conform to the IteratorResult interface. Invoking this method notifies the Iterator object that the caller does not intend to make any more next

"any more" next implies there was at least one. Does that convince you ?

It is perhaps never stated explicitly, but this is how they implemented generator functions. This is the heuristic used by:
* for .. of
* array spread
* etc.

If someone implements things differently it is already broken using js primitives.

Reply:
I like your line of thought, but in fact you are disproved by the array spread operator, which will call return even in the unusual case that it does not attempt to spread any items.

https://jsfiddle.net/dyqp54vr/

Reply:
Alternative idea. How are you going to call "return" in the instance? if the constructor fails you don't get the instance at all.

Reply:
I'm just saying the code should be:
```js
let iterator
...
function * generateGroups() {
  try {
    iterator = iterable[Symbol.iterator]()
  } finally {}
}
```
I don't know what you're talking about in terms of the constructor failing.

Reply:
with the empty finally block ?

Reply:
Uh, no? Keep everything the way it is now. Just move the initialization inside finally. Or don't, and I'll open a PR once this is merged.